src/coreclr/nativeaot/Runtime/GCHelpers.cpp

```diff
@@ -550,6 +619,19 @@ static Object* GcAllocInternal(MethodTable* pEEType, uint32_t uFlags, uintptr_t
         ((Array*)pObject)->InitArrayLength((uint32_t)numElements);
     }
 
+    if (isSampled)
+    {
+        FireAllocationSampled((GC_ALLOC_FLAGS)uFlags, cbSize, samplingBudget, pObject);
+    }
+
+    // There are a variety of conditions that may have invalidated the previous combined_limit value
```

Comment:
Do we need to account for unused sampling quota from before the last combined limit reset?

The scenario that I am thinking about: One thread allocates a lot and triggers the GC a lot, second thread allocates a little. The second thread allocates a little. The second thread should get its fair share of samples, but I do not think it is going to the case since its sampling quota will be reset by GC running before it gets a chance to report a sample most of the time.

Reply:
> Do we need to account for unused sampling quota from before the last combined limit reset?

To the best of my understanding, discarding the remainder of sampling quota and regenerating the limit does not cause any problem for statistical accuracy as long as the decision to discard is independent of the sample limit value. There is some additional discussion of this late in the feature doc: https://github.com/dotnet/runtime/pull/104955/files#diff-f0760b2f7659c7433b5313f11d30bf17f0061252058af8511ad117bfee99ceb3R310

> but I do not think it is going to the case with this implementation since its sampling quota will be reset by GC running before it gets a chance to report a sample most of the time.

The sampling quota is generated in such a way that every allocated byte has a 1/102,400 chance of being sampled. For example if thread A has only allocated 20 bytes between each GC then it will have  ~20/102,400 chance of sampling at least one byte during that time. That is a small chance, but it is the proportionally correct chance based on the amount of allocating this thread performed.